# Обработка данных

In [1]:
import pandas as pd
import numpy as np
features = pd.read_csv(r'S:\Coursera\Введение в машинное обучение\Итоговое задание\features.csv', index_col='match_id')

In [2]:
pd.set_option('display.max_columns', None)
features

start_time  lobby_type  r1_hero  r1_level  r1_xp  r1_gold  r1_lh  \
match_id                                                                     
0         1430198770           7       11         5   2098     1489     20   
1         1430220345           0       42         4   1188     1033      9   
2         1430227081           7       33         4   1319     1270     22   
3         1430263531           1       29         4   1779     1056     14   
4         1430282290           7       13         4   1431     1090      8   
...              ...         ...      ...       ...    ...      ...    ...   
114402    1450265551           1       47         4   1706     1198     17   
114403    1450277704           0       43         4   1793     1416     17   
114404    1450291848           1       98         4   1399      540      1   
114405    1450292986           1      100         3   1135      766      6   
114406    1450313370           7       50         3   1053      799      7   

          r1_kills  r1_deaths  r1_items  r2_hero  r2_level  r2_xp  r2_gold  \
match_id                                                                     
0                0          0         7       67         3    842      991   
1                0          1        12       49         4   1596      993   
2                0          0        12       98         3   1314      775   
3                0          0         5       30         2    539      539   
4                1          0         8       27         2    629      552   
...            ...        ...       ...      ...       ...    ...      ...   
114402           0          1         8        7         2    616      535   
114403           0          1         5       26         3    764      800   
114404           0          0         5       11         4   1448     1371   
114405           0          2         6       72         5   1954     1457   
114406           0          0         7       19         5   2097     1818   

          r2_lh  r2_kills  r2_deaths  r2_items  r3_hero  r3_level  r3_xp  \
match_id                                                                   
0            10         0          0         4       29         5   1909   
1            10         0          1         7       67         4   1506   
2             6         0          0         6       20         3   1297   
3             1         0          0         6       75         5   2037   
4             0         0          1         7       30         3    884   
...         ...       ...        ...       ...      ...       ...    ...   
114402        0         0          2        10        1         3    879   
114403        0         1          0         9        4         3   1130   
114404       17         0          1        13      112         3    658   
114405       22         0          0         9       79         2    471   
114406       23         1          2         9       84         3   1103   

          r3_gold  r3_lh  r3_kills  r3_deaths  r3_items  r4_hero  r4_level  \
match_id                                                                     
0            1143     10         0          0         8       20         3   
1            1502     18         1          0         7       37         3   
2             909      0         1          0         6       27         5   
3            1139     15         0          0         6       37         2   
4             927      0         1          0         8       72         3   
...           ...    ...       ...        ...       ...      ...       ...   
114402       1798     22         1          1         9       21         4   
114403       1585     20         1          1        14       29         4   
114404        646      2         0          0        10       81         4   
114405        634      2         0          1        10       20         1   
114406        889      3         1          0         9

Выделим в списки категориальные и числовые признаки:
(порядковые признаки - уровни героев - я отнесла к числовым).

In [3]:
#колонки с категориальными признаками
cat_cols = ['lobby_type',
           'r1_hero',
           'r2_hero',
           'r3_hero',
           'r4_hero',
           'r5_hero',
           'd1_hero',
           'd2_hero',
           'd3_hero',
           'd4_hero',
           'd5_hero',
           'first_blood_player1',
           'first_blood_player2'
           ]

In [4]:
#колонки с числовыми признаками
num_cols = ['r1_level',
           'r1_xp',
           'r1_gold',
           'r1_lh',
           'r1_kills',
           'r1_deaths',
           'r1_items',
           'r2_level',
           'r2_xp',
           'r2_gold',
           'r2_lh',
           'r2_kills',
           'r2_deaths',
           'r2_items',
           'r3_level',
           'r3_xp',
           'r3_gold',
           'r3_lh',
           'r3_kills',
           'r3_deaths',
           'r3_items',
           'r4_level',
           'r4_xp',
           'r4_gold',
           'r4_lh',
           'r4_kills',
           'r4_deaths',
           'r4_items',
           'r5_level',
           'r5_xp',
           'r5_gold',
           'r5_lh',
           'r5_kills',
           'r5_deaths',
           'r5_items',
           'd1_level',
           'd1_xp',
           'd1_gold',
           'd1_lh',
           'd1_kills',
           'd1_deaths',
           'd1_items',
           'd2_level',
           'd2_xp',
           'd2_gold',
           'd2_lh',
           'd2_kills',
           'd2_deaths',
           'd2_items',
           'd3_level',
           'd3_xp',
           'd3_gold',
           'd3_lh',
           'd3_kills',
           'd3_deaths',
           'd3_items',
           'd4_level',
           'd4_xp',
           'd4_gold',
           'd4_lh',
           'd4_kills',
           'd4_deaths',
           'd4_items',
           'd5_level',
           'd5_xp',
           'd5_gold',
           'd5_lh',
           'd5_kills',
           'd5_deaths',
           'd5_items',
           'first_blood_time',
           'first_blood_team',
           'radiant_bottle_time',
           'radiant_courier_time',
           'radiant_flying_courier_time',
           'radiant_tpscroll_count',
           'radiant_boots_count',
           'radiant_ward_observer_count',
           'radiant_ward_sentry_count',
           'radiant_first_ward_time',
           'dire_bottle_time',
           'dire_courier_time',
           'dire_flying_courier_time',
           'dire_tpscroll_count',
           'dire_boots_count',
           'dire_ward_observer_count',
           'dire_ward_sentry_count',
           'dire_first_ward_time'
           ]

#колонки с категориальными и числовыми признаками
feature_cols = num_cols + cat_cols

#целевая переменная
target_col = ['radiant_win']


2. Как называется столбец, содержащий целевую переменную?

'radiant_win'


In [5]:
#функция выделяет из массива данных X и y

def create_y_X(data):
    y = data[target_col]
    X = data[feature_cols]

    return y, X

In [6]:
y, X = create_y_X(features)

In [7]:
X.corr().style.set_precision(2)

#В матрице корреляций можно увидеть достаточно большую корреляцию между признаками уровня, 
#опыта, ценности героя и числом убитых юнитов. В дальнейшем это можно использовать для
#создания новых признаков.


1. Какие признаки имеют пропуски среди своих значений? Что могут означать пропуски в этих признаках (ответьте на этот вопрос для двух любых признаков)?

Пропуски связаны с тем, что событие (приобретение предмета или первая кровь) не произошло.


In [8]:
#Выявление столбцов с пропусками в данных. 

print('Пропуски данных в столбцах:')
for i in feature_cols:
    if X[i].count() != 97230:
        print(i, X[i].count())

Пропуски данных в столбцах:
first_blood_time 77677
first_blood_team 77677
radiant_bottle_time 81539
radiant_courier_time 96538
radiant_flying_courier_time 69751
radiant_first_ward_time 95394
dire_bottle_time 81087
dire_courier_time 96554
dire_flying_courier_time 71132
dire_first_ward_time 95404
first_blood_player1 77677
first_blood_player2 53243


In [9]:
#Выделяем колонки, имеющие пропуски в данных и относящиеся ко времени 

time_gap_cols = ['first_blood_time', 
                 'radiant_bottle_time', 'radiant_courier_time', 'radiant_flying_courier_time', 'radiant_first_ward_time',
                 'dire_bottle_time', 'dire_courier_time', 'dire_flying_courier_time', 'dire_first_ward_time']

In [10]:
#Данные есть о первых 5 минутах матча, значит максимально возможное значение в таких колонках будет 5*60 = 300 сек. Проверим:

features['dire_first_ward_time'].max()

300.0

Обработка пропусков: 
1. first_blood_player заменяем на 0, так как кодировка игроков начинается с 1. 
2. first_blood_team заменяем на 0.5, что будет значить, что ни 0 ни 1 команда не совершила first blood 
3. для колонок time_gap_cols выбираем значение, сильно больше максимального, что будет значить, что событие произошло позже данного для исследования времени (300 сек)

In [11]:
X['first_blood_player1'] = X['first_blood_player1'].fillna(0)
X['first_blood_player2'] = X['first_blood_player2'].fillna(0)
X['first_blood_team'] = X['first_blood_team'].fillna(.5)
X[time_gap_cols] = X[time_gap_cols].fillna(1000)

<ipython-input-11-76f6209d5166>:7: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  X['first_blood_player1'] = X['first_blood_player1'].fillna(0)
<ipython-input-11-76f6209d5166>:8: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  X['first_blood_player2'] = X['first_blood_player2'].fillna(0)
<ipython-input-11-76f6209d5166>:9: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http

In [12]:
X

r1_level  r1_xp  r1_gold  r1_lh  r1_kills  r1_deaths  r1_items  \
match_id                                                                   
0                5   2098     1489     20         0          0         7   
1                4   1188     1033      9         0          1        12   
2                4   1319     1270     22         0          0        12   
3                4   1779     1056     14         0          0         5   
4                4   1431     1090      8         1          0         8   
...            ...    ...      ...    ...       ...        ...       ...   
114402           4   1706     1198     17         0          1         8   
114403           4   1793     1416     17         0          1         5   
114404           4   1399      540      1         0          0         5   
114405           3   1135      766      6         0          2         6   
114406           3   1053      799      7         0          0         7   

          r2_level  r2_xp  r2_gold  r2_lh  r2_kills  r2_deaths  r2_items  \
match_id                                                                   
0                3    842      991     10         0          0         4   
1                4   1596      993     10         0          1         7   
2                3   1314      775      6         0          0         6   
3                2    539      539      1         0          0         6   
4                2    629      552      0         0          1         7   
...            ...    ...      ...    ...       ...        ...       ...   
114402           2    616      535      0         0          2        10   
114403           3    764      800      0         1          0         9   
114404           4   1448     1371     17         0          1        13   
114405           5   1954     1457     22         0          0         9   
114406           5   2097     1818     23         1          2         9   

          r3_level  r3_xp  r3_gold  r3_lh  r3_kills  r3_deaths  r3_items  \
match_id                                                                   
0                5   1909     1143     10         0          0         8   
1                4   1506     1502     18         1          0         7   
2                3   1297      909      0         1          0         6   
3                5   2037     1139     15         0          0         6   
4                3    884      927      0         1          0         8   
...            ...    ...      ...    ...       ...        ...       ...   
114402           3    879     1798     22         1          1         9   
114403           3   1130     1585     20         1          1        14   
114404           3    658      646      2         0          0        10   
114405           2    471      634      2         0          1        10   
114406           3   1103      889      3         1          0         9   

          r4_level  r4_xp  r4_gold  r4_lh  r4_kills  r4_deaths  r4_items  \
match_id                                                                   
0                3    757      741      6         0          0         7   
1                3    669      631      7         0          0         7   
2                5   2360     2096     26         1          1         6   
3                2    591      499      0         0          0         6   
4                3    925     1439     16         1          0        11   
...            ...    ...      ...    ...       ...        ...       ...   
114402           4   1835     1350     11         1          1         9   
114403           4   1546     1166      8         1          0         6   
114404           4   1710     1833     24         1          0        10   
114405           1    265      605      1         0          0         9   
114406           4   1293     1675     11         1          0        11   

          r5_level  r5_xp  r5_gold  r5_lh  r5_kills  r

# Градиентный бустинг 
## 1. Градиентный бустинг в лоб
Игнорируем категориальные фичи

In [13]:
from sklearn.ensemble import GradientBoostingClassifier

In [16]:
from sklearn.model_selection import train_test_split

In [23]:
from sklearn.metrics import roc_auc_score

In [14]:
y = np.array(y)
y = y.ravel()

In [19]:
X_train, X_test, y_train, y_test = train_test_split(X, y, train_size=0.8, random_state=241)
gbc = GradientBoostingClassifier(random_state=241)

In [20]:
gbc.fit(X_train, y_train)

GradientBoostingClassifier(random_state=241)

In [24]:
Pred_gbc_test = gbc.predict_proba(X_test)
Pred_gbc_train = gbc.predict_proba(X_train)
roc_auc_test = roc_auc_score(y_test, Pred_gbc_test[:,1])
roc_auc_train = roc_auc_score(y_train, Pred_gbc_train[:,1])
print('значение roc_auc на test: ', roc_auc_test)
print('значение roc_auc на train: ', roc_auc_train)
print('среднее значение roc_auc: ', (roc_auc_test + roc_auc_train)/2)

значение roc_auc на test:  0.7112372423391344
значение roc_auc на train:  0.7250516789760564
среднее значение roc_auc:  0.7181444606575954


Вывод: качество низкое, подход неправильный (нельзя игнорировать категориальные фичи)

## 2. Улучшенный градиентный бустинг

In [36]:
from sklearn.ensemble import GradientBoostingClassifier
from sklearn.model_selection import train_test_split, KFold, GridSearchCV
from sklearn.metrics import roc_auc_score

In [27]:
from sklearn.model_selection import cross_val_score

In [26]:
#one-hot-encoding: обработка кат. признаков
X = pd.get_dummies(X, columns=cat_cols, drop_first=True)
X

r1_level  r1_xp  r1_gold  r1_lh  r1_kills  r1_deaths  r1_items  \
match_id                                                                   
0                5   2098     1489     20         0          0         7   
1                4   1188     1033      9         0          1        12   
2                4   1319     1270     22         0          0        12   
3                4   1779     1056     14         0          0         5   
4                4   1431     1090      8         1          0         8   
...            ...    ...      ...    ...       ...        ...       ...   
114402           4   1706     1198     17         0          1         8   
114403           4   1793     1416     17         0          1         5   
114404           4   1399      540      1         0          0         5   
114405           3   1135      766      6         0          2         6   
114406           3   1053      799      7         0          0         7   

          r2_level  r2_xp  r2_gold  r2_lh  r2_kills  r2_deaths  r2_items  \
match_id                                                                   
0                3    842      991     10         0          0         4   
1                4   1596      993     10         0          1         7   
2                3   1314      775      6         0          0         6   
3                2    539      539      1         0          0         6   
4                2    629      552      0         0          1         7   
...            ...    ...      ...    ...       ...        ...       ...   
114402           2    616      535      0         0          2        10   
114403           3    764      800      0         1          0         9   
114404           4   1448     1371     17         0          1        13   
114405           5   1954     1457     22         0          0         9   
114406           5   2097     1818     23         1          2         9   

          r3_level  r3_xp  r3_gold  r3_lh  r3_kills  r3_deaths  r3_items  \
match_id                                                                   
0                5   1909     1143     10         0          0         8   
1                4   1506     1502     18         1          0         7   
2                3   1297      909      0         1          0         6   
3                5   2037     1139     15         0          0         6   
4                3    884      927      0         1          0         8   
...            ...    ...      ...    ...       ...        ...       ...   
114402           3    879     1798     22         1          1         9   
114403           3   1130     1585     20         1          1        14   
114404           3    658      646      2         0          0        10   
114405           2    471      634      2         0          1        10   
114406           3   1103      889      3         1          0         9   

          r4_level  r4_xp  r4_gold  r4_lh  r4_kills  r4_deaths  r4_items  \
match_id                                                                   
0                3    757      741      6         0          0         7   
1                3    669      631      7         0          0         7   
2                5   2360     2096     26         1          1         6   
3                2    591      499      0         0          0         6   
4                3    925     1439     16         1          0        11   
...            ...    ...      ...    ...       ...        ...       ...   
114402           4   1835     1350     11         1          1         9   
114403           4   1546     1166      8         1          0         6   
114404           4   1710     1833     24         1          0        10   
114405           1    265      605      1         0          0         9   
114406           4   1293     1675     11         1          0        11   

          r5_level  r5_xp  r5_gold  r5_lh  r5_kills  r

In [32]:
gbc = GradientBoostingClassifier(random_state=241)

In [33]:
X_train, X_test, y_train, y_test = train_test_split(X, y, train_size=0.8, random_state=241)

In [34]:
#сетка гиперпараметров для поиска лучших
params = {'n_estimators': (10, 20, 30, 50),
             'max_depth': (3, 5, 7, 10, 20)}

In [37]:
#ищем по сетке
grid = GridSearchCV(gbc, params, cv=5)
grid.fit(X_train, y_train)

GridSearchCV(cv=5, estimator=GradientBoostingClassifier(random_state=241),
             param_grid={'max_depth': (3, 5, 7, 10, 20),
                         'n_estimators': (10, 20, 30, 50)})

In [38]:
grid.best_params_

{'max_depth': 7, 'n_estimators': 50}

4. Имеет ли смысл использовать больше 30 деревьев в градиентном бустинге? Что бы вы предложили делать, чтобы ускорить его обучение при увеличении количества деревьев?

Есть смысл использовать более 30 деревьев. Как видно ниже, точность при 30 деревьях 0.7029, при 50 деревьях 0.7135.
Чтобы ускорить обучение, можно использовать для обучения и кросс-валидации не всю выборку, а некоторое ее подмножество — например, половину объектов.

In [39]:
#Обучаем модель с найденными параметрами:

gbc_best = GradientBoostingClassifier(n_estimators=50, max_depth=7, random_state=241)
gbc_best.fit(X_train, y_train)

GradientBoostingClassifier(max_depth=7, n_estimators=50, random_state=241)

In [44]:
Pred_gbc_best_test = gbc_best.predict_proba(X_test)
Pred_gbc_best_train = gbc_best.predict_proba(X_train)
roc_auc_test = roc_auc_score(y_test, Pred_gbc_best_test[:,1])
roc_auc_train = roc_auc_score(y_train, Pred_gbc_best_train[:,1])
print('значение roc_auc на test: ', round(roc_auc_test, 4))
print('значение roc_auc на train: ', round(roc_auc_train, 4))
print('среднее значение roc_auc: ', round((roc_auc_test + roc_auc_train)/2, 4))

значение roc_auc на test:  0.7135
значение roc_auc на train:  0.8108
среднее значение roc_auc:  0.7621


Результаты gbc_best на каггле: 0.71048 (не густо)

Наблюдается значительное прееобучение

3. Как долго проводилась кросс-валидация для градиентного бустинга с 30 деревьями? Инструкцию по измерению времени можно найти ниже по тексту. Какое качество при этом получилось? Напомним, что в данном задании мы используем метрику качества AUC-ROC.

In [46]:
#кросс-валидация для градиентного бустинга с 30 деревьями:

import time
import datetime

start_time = datetime.datetime.now()

gbc_30 = GradientBoostingClassifier(n_estimators=30, max_depth=7, random_state=241)
kf = KFold(n_splits=5, shuffle=True)

quality = cross_val_score(gbc_30, X, y, cv=kf, scoring='roc_auc') #качество на кросс-валидации
quality_mean = np.mean(quality) #среднее качество на кросс-валидации

print('значения roc_auc на кросс-валидации:', quality)
print('среднее значение roc_auc:', quality_mean)

print('Time elapsed:', datetime.datetime.now() - start_time)

значения roc_auc на кросс-валидации: [0.70360797 0.70438706 0.70643998 0.70339944 0.69656549]
среднее значение roc_auc: 0.7028799849210003
Time elapsed: 0:14:33.080399


## 3. CatBoost

In [64]:
!pip install catboost
import catboost

In [65]:
from sklearn.model_selection import train_test_split, GridSearchCV
from sklearn.metrics import roc_auc_score

In [66]:
y, X = create_y_X(features)
X['first_blood_player1'] = X['first_blood_player1'].fillna(0)
X['first_blood_player2'] = X['first_blood_player2'].fillna(0)
X['first_blood_team'] = X['first_blood_team'].fillna(.5)
X[time_gap_cols] = X[time_gap_cols].fillna(1000)

<ipython-input-66-0bba7304d03a>:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  X['first_blood_player1'] = X['first_blood_player1'].fillna(0)
<ipython-input-66-0bba7304d03a>:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  X['first_blood_player2'] = X['first_blood_player2'].fillna(0)
<ipython-input-66-0bba7304d03a>:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http

In [67]:
#Сейчас некоторые кат. признаки типа float. Напишем функцию, преобразующую их в int

def cat_to_int(data, cat_cols):
    data[cat_cols] = data[cat_cols].astype('int64')
    return data

In [68]:
#Создадим матрицу признаков для кэтбустинга

X_catboost = cat_to_int(X, cat_cols)

In [69]:
#дефолтный кэтбустинг
cat = catboost.CatBoostClassifier(eval_metric='AUC', silent=True, random_state=241, cat_features=cat_cols)

X_catboost_train, X_catboost_test, y_train, y_test = train_test_split(X_catboost, y, train_size=0.8, random_state=241)


In [72]:
#обучение
import time
import datetime

start_time = datetime.datetime.now()

cat.fit(X_catboost_train, y_train)

print('Time elapsed:', datetime.datetime.now() - start_time)

Time elapsed: 0:01:39.067833


In [73]:
#предсказание

Pred_catboost_test = cat.predict_proba(X_catboost_test)
Pred_catboost_train = cat.predict_proba(X_catboost_train)
roc_auc_test = roc_auc_score(y_test, Pred_catboost_test[:,1])
roc_auc_train = roc_auc_score(y_train, Pred_catboost_train[:,1])
print('значение roc_auc на test: ', roc_auc_test)
print('значение roc_auc на train: ', roc_auc_train)
print('среднее значение roc_auc: ', (roc_auc_test + roc_auc_train)/2)

значение roc_auc на test:  0.7351104250565577
значение roc_auc на train:  0.8177808303104578
среднее значение roc_auc:  0.7764456276835077


Вывод: 
1. Наблюдается значительное переобучение 
2. Результат на каггле 0.73055
3. Время исполнения кода значительно ниже, чем при использовании GradientBoostingClassifier (1 мин 39 сек   vs   14 мин 33 сек)
4. Качество выше на 2.5%, чем при использовании GradientBoostingClassifier (0.7351    vs    0.7029)

# ПРЕДСКАЗАНИЯ ДЛЯ ГРАДИЕНТНЫХ БУСТИНГОВ

In [58]:
best_model = gbc_best

In [55]:
X_test_test = pd.read_csv(r'S:\Coursera\Введение в машинное обучение\Итоговое задание\features_test.csv', index_col='match_id')
X_test_test = X_test_test[feature_cols]
X_test_test['first_blood_player1'] = X_test_test['first_blood_player1'].fillna(0)
X_test_test['first_blood_player2'] = X_test_test['first_blood_player2'].fillna(0)
X_test_test['first_blood_team'] = X_test_test['first_blood_team'].fillna(.5)
X_test_test[time_gap_cols] = X_test_test[time_gap_cols].fillna(1000)
X_test_test = pd.get_dummies(X_test_test, columns=cat_cols, drop_first=True) #для случайного леса
#X_test_test = cat_to_int(X_test_test, cat_cols) #для кэтбуста

X_test_test

r1_level  r1_xp  r1_gold  r1_lh  r1_kills  r1_deaths  r1_items  \
match_id                                                                   
6                4   1103     1089      8         0          1         9   
7                2    556      570      1         0          0         9   
10               2    751      808      1         0          0        13   
13               3    708      903      1         1          1        11   
16               4   1259      661      4         0          0         9   
...            ...    ...      ...    ...       ...        ...       ...   
114369           5   2054     1941     27         0          1         8   
114377           3    748      605      1         0          0        12   
114378           2    575      499      0         0          0         8   
114393           4   1844     1176      8         1          2         8   
114398           4   1215     1305     13         0          0        10   

          r2_level  r2_xp  r2_gold  r2_lh  r2_kills  r2_deaths  r2_items  \
match_id                                                                   
6                3   1183      963      9         0          2         9   
7                4   1194     1386     21         0          0         7   
10               2    421      569      3         0          0         6   
13               2    672      901      6         0          0         8   
16               5   1703      964     10         0          1         6   
...            ...    ...      ...    ...       ...        ...       ...   
114369           4   1460      936      7         0          1         8   
114377           3   1130     1385     16         1          0         9   
114378           3    697      686      3         0          0         8   
114393           3   1013      841      6         0          0         7   
114398           5   1867     1573     19         1          0         9   

          r3_level  r3_xp  r3_gold  r3_lh  r3_kills  r3_deaths  r3_items  \
match_id                                                                   
6                5   2289     2043     20         2          0         6   
7                4   1374     1064     13         0          1         8   
10               3    960      896      3         1          2         6   
13               5   2141     1626     28         0          0         7   
16               3   1099      942      2         1          0         6   
...            ...    ...      ...    ...       ...        ...       ...   
114369           3   1208     1579     22         0          1        13   
114377           3    779      660      0         0          0         8   
114378           5   2050      980      8         0          1         8   
114393           5   1920     1260     16         0          0         6   
114398           4   1405      954      6         0          1        10   

          r4_level  r4_xp  r4_gold  r4_lh  r4_kills  r4_deaths  r4_items  \
match_id                                                                   
6                3    917     1006     14         0          0         6   
7                2    774      737      0         1          0         8   
10               5   1877     1511     23         0          0         8   
13               4   1579     2169     28         2          0         7   
16               3   1285     1013     10         0          0         8   
...            ...    ...      ...    ...       ...        ...       ...   
114369           2    464      816      1         0          1        10   
114377           4   1463     1362     18         0          0         8   
114378           3    853     1014     13         0          1         6   
114393           3    626      500      0         0          0         8   
114398           3    979     1064      2         2          0        11   

          r5_level  r5_xp  r5_gold  r5_lh  r5_kills  r

In [59]:
submission = pd.read_csv(r'S:\Coursera\Введение в машинное обучение\Итоговое задание\submission.csv')

In [60]:
submission['match_id'] = X_test_test.index

In [61]:
submission['radiant_win'] = best_model.predict_proba(X_test_test)[:,1]

In [63]:
submission.to_csv(r'S:\Coursera\Введение в машинное обучение\Итоговое задание\my_submission_gbc_best.csv', index=False)

# Логистическая регрессия

In [74]:
#Готовим данные для лог. регрессии заново и немного по-другому
y, X = create_y_X(features)
X['first_blood_player1'] = X['first_blood_player1'].fillna(0)
X['first_blood_player2'] = X['first_blood_player2'].fillna(0)
X['first_blood_team'] = X['first_blood_team'].fillna(.5)
X[time_gap_cols] = X[time_gap_cols].fillna(1000)
#one-hot-encoding:
X = pd.get_dummies(X, columns=cat_cols, drop_first=True)
X

<ipython-input-74-54abf3011955>:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  X['first_blood_player1'] = X['first_blood_player1'].fillna(0)
<ipython-input-74-54abf3011955>:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  X['first_blood_player2'] = X['first_blood_player2'].fillna(0)
<ipython-input-74-54abf3011955>:5: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http

r1_level  r1_xp  r1_gold  r1_lh  r1_kills  r1_deaths  r1_items  \
match_id                                                                   
0                5   2098     1489     20         0          0         7   
1                4   1188     1033      9         0          1        12   
2                4   1319     1270     22         0          0        12   
3                4   1779     1056     14         0          0         5   
4                4   1431     1090      8         1          0         8   
...            ...    ...      ...    ...       ...        ...       ...   
114402           4   1706     1198     17         0          1         8   
114403           4   1793     1416     17         0          1         5   
114404           4   1399      540      1         0          0         5   
114405           3   1135      766      6         0          2         6   
114406           3   1053      799      7         0          0         7   

          r2_level  r2_xp  r2_gold  r2_lh  r2_kills  r2_deaths  r2_items  \
match_id                                                                   
0                3    842      991     10         0          0         4   
1                4   1596      993     10         0          1         7   
2                3   1314      775      6         0          0         6   
3                2    539      539      1         0          0         6   
4                2    629      552      0         0          1         7   
...            ...    ...      ...    ...       ...        ...       ...   
114402           2    616      535      0         0          2        10   
114403           3    764      800      0         1          0         9   
114404           4   1448     1371     17         0          1        13   
114405           5   1954     1457     22         0          0         9   
114406           5   2097     1818     23         1          2         9   

          r3_level  r3_xp  r3_gold  r3_lh  r3_kills  r3_deaths  r3_items  \
match_id                                                                   
0                5   1909     1143     10         0          0         8   
1                4   1506     1502     18         1          0         7   
2                3   1297      909      0         1          0         6   
3                5   2037     1139     15         0          0         6   
4                3    884      927      0         1          0         8   
...            ...    ...      ...    ...       ...        ...       ...   
114402           3    879     1798     22         1          1         9   
114403           3   1130     1585     20         1          1        14   
114404           3    658      646      2         0          0        10   
114405           2    471      634      2         0          1        10   
114406           3   1103      889      3         1          0         9   

          r4_level  r4_xp  r4_gold  r4_lh  r4_kills  r4_deaths  r4_items  \
match_id                                                                   
0                3    757      741      6         0          0         7   
1                3    669      631      7         0          0         7   
2                5   2360     2096     26         1          1         6   
3                2    591      499      0         0          0         6   
4                3    925     1439     16         1          0        11   
...            ...    ...      ...    ...       ...        ...       ...   
114402           4   1835     1350     11         1          1         9   
114403           4   1546     1166      8         1          0         6   
114404           4   1710     1833     24         1          0        10   
114405           1    265      605      1         0          0         9   
114406           4   1293     1675     11         1          0        11   

          r5_level  r5_xp  r5_gold  r5_lh  r5_kills  r

In [94]:
y = np.array(y)
y = y.ravel()

In [95]:
from sklearn.preprocessing import StandardScaler
from sklearn.linear_model import LogisticRegression, LogisticRegressionCV
from sklearn.model_selection import train_test_split, GridSearchCV
from sklearn.metrics import roc_auc_score

In [96]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=241)

In [97]:
#масштабируем признаки

pca = StandardScaler()
X_train_ = pca.fit_transform(X_train)
X_test_ = pca.fit_transform(X_test)

In [107]:
X_ = pca.fit_transform(X)

In [98]:
#дефолтная лог. регрессия

logreg = LogisticRegression(penalty='l2', random_state=241, max_iter=500)
logreg.fit(X_train_, y_train)

LogisticRegression(max_iter=500, random_state=241)

In [99]:
Pred_logreg_test = logreg.predict_proba(X_test_)
Pred_logreg_train = logreg.predict_proba(X_train_)
roc_auc_test = roc_auc_score(y_test, Pred_logreg_test[:,1])
roc_auc_train = roc_auc_score(y_train, Pred_logreg_train[:,1])
print('значение roc_auc на test: ', round(roc_auc_test, 4))
print('значение roc_auc на train: ', round(roc_auc_train, 4))
print('среднее значение roc_auc: ', round((roc_auc_test + roc_auc_train)/2, 4))

значение roc_auc на test:  0.7461
значение roc_auc на train:  0.7619
среднее значение roc_auc:  0.754


Вывод: 
    1. Переобучение незначительно 
    2. roc_auc выше, чем у кэтбустинга (0.7461 vs 0.7351)

# Логистическая регрессия с поиском лучших параметров

In [101]:
# Значения коэффициента С: 100 значений в промежутке от 0.001 до 1000

a = np.linspace(0.001, 1000, 100)
logregcv = LogisticRegressionCV(Cs=a , fit_intercept=True, cv=5, scoring='roc_auc', random_state=241, refit=True, max_iter=500)

In [102]:
logregcv.fit(X_train_, y_train)

LogisticRegressionCV(Cs=array([1.00000e-03, 1.01020e+01, 2.02030e+01, 3.03040e+01, 4.04050e+01,
       5.05060e+01, 6.06070e+01, 7.07080e+01, 8.08090e+01, 9.09100e+01,
       1.01011e+02, 1.11112e+02, 1.21213e+02, 1.31314e+02, 1.41415e+02,
       1.51516e+02, 1.61617e+02, 1.71718e+02, 1.81819e+02, 1.91920e+02,
       2.02021e+02, 2.12122e+02, 2.22223e+02, 2.32324e+02, 2.42425e+02,
       2.52526e+02, 2.62627...
       7.57576e+02, 7.67677e+02, 7.77778e+02, 7.87879e+02, 7.97980e+02,
       8.08081e+02, 8.18182e+02, 8.28283e+02, 8.38384e+02, 8.48485e+02,
       8.58586e+02, 8.68687e+02, 8.78788e+02, 8.88889e+02, 8.98990e+02,
       9.09091e+02, 9.19192e+02, 9.29293e+02, 9.39394e+02, 9.49495e+02,
       9.59596e+02, 9.69697e+02, 9.79798e+02, 9.89899e+02, 1.00000e+03]),
                     cv=5, max_iter=500, random_state=241, scoring='roc_auc')

In [103]:
Pred_logregcv_test = logregcv.predict_proba(X_test_)
Pred_logregcv_train = logregcv.predict_proba(X_train_)
roc_auc_test = roc_auc_score(y_test, Pred_logregcv_test[:,1])
roc_auc_train = roc_auc_score(y_train, Pred_logregcv_train[:,1])
print('значение roc_auc на test: ', round(roc_auc_test, 4))
print('значение roc_auc на train: ', round(roc_auc_train, 4))
print('среднее значение roc_auc: ', round((roc_auc_test + roc_auc_train)/2, 4))

значение roc_auc на test:  0.7461
значение roc_auc на train:  0.7619
среднее значение roc_auc:  0.754


Качество не изменилось

In [104]:
#Лучший коэффициент с

c_best = logregcv.C_
c_best = np.max(c_best)
c_best

60.607000000000006

In [108]:
#кросс-валидация для лог. регрессии, С = 60.607000000000006:

import time
import datetime

start_time = datetime.datetime.now()

logreg_best = LogisticRegression(C=c_best, penalty='l2', max_iter=500, random_state=241)
kf = KFold(n_splits=5, shuffle=True)

quality = cross_val_score(logreg_best, X_, y, cv=kf, scoring='roc_auc') #качество на кросс-валидации
quality_mean = np.mean(quality) #среднее качество на кросс-валидации

print('значения roc_auc на кросс-валидации:', quality)
print('среднее значение roc_auc:', quality_mean)

print('Time elapsed:', datetime.datetime.now() - start_time)

значения roc_auc на кросс-валидации: [0.74366636 0.7486564  0.74708533 0.74447908 0.74120363]
среднее значение roc_auc: 0.7450181610812278
Time elapsed: 0:00:51.775708


In [109]:
logreg_best.fit(X_, y)

LogisticRegression(C=60.607000000000006, max_iter=500, random_state=241)

Вывод: Результат на каггле 0.75053 - лучшее качество из всех моделей

1. Какое качество получилось у логистической регрессии над всеми исходными признаками? Как оно соотносится с качеством градиентного бустинга? Чем вы можете объяснить эту разницу? Быстрее ли работает логистическая регрессия по сравнению с градиентным бустингом?
Качество у логистической регрессии 0.75053, у град. бустинга 0.7351.
Логистическая регрессия работает быстрее по сравнению с градиентным бустингом (52 сек   vs   1 мин 39 сек).
Причиной, по которой лог. регрессия работает на этом датасете лучше, является разреженность матрицы признаков.

2. Как влияет на качество логистической регрессии удаление категориальных признаков (укажите новое значение метрики качества)? Чем вы можете объяснить это изменение?
См. код ниже
удаление категориальных признаков снижает значение roc_auc = 0.7211

In [114]:
y, X = create_y_X(features)
X['first_blood_player1'] = X['first_blood_player1'].fillna(0)
X['first_blood_player2'] = X['first_blood_player2'].fillna(0)
X['first_blood_team'] = X['first_blood_team'].fillna(.5)
X[time_gap_cols] = X[time_gap_cols].fillna(1000)
X = X[num_cols]
X

<ipython-input-114-92e7da31768c>:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  X['first_blood_player1'] = X['first_blood_player1'].fillna(0)
<ipython-input-114-92e7da31768c>:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  X['first_blood_player2'] = X['first_blood_player2'].fillna(0)
<ipython-input-114-92e7da31768c>:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: h

r1_level  r1_xp  r1_gold  r1_lh  r1_kills  r1_deaths  r1_items  \
match_id                                                                   
0                5   2098     1489     20         0          0         7   
1                4   1188     1033      9         0          1        12   
2                4   1319     1270     22         0          0        12   
3                4   1779     1056     14         0          0         5   
4                4   1431     1090      8         1          0         8   
...            ...    ...      ...    ...       ...        ...       ...   
114402           4   1706     1198     17         0          1         8   
114403           4   1793     1416     17         0          1         5   
114404           4   1399      540      1         0          0         5   
114405           3   1135      766      6         0          2         6   
114406           3   1053      799      7         0          0         7   

          r2_level  r2_xp  r2_gold  r2_lh  r2_kills  r2_deaths  r2_items  \
match_id                                                                   
0                3    842      991     10         0          0         4   
1                4   1596      993     10         0          1         7   
2                3   1314      775      6         0          0         6   
3                2    539      539      1         0          0         6   
4                2    629      552      0         0          1         7   
...            ...    ...      ...    ...       ...        ...       ...   
114402           2    616      535      0         0          2        10   
114403           3    764      800      0         1          0         9   
114404           4   1448     1371     17         0          1        13   
114405           5   1954     1457     22         0          0         9   
114406           5   2097     1818     23         1          2         9   

          r3_level  r3_xp  r3_gold  r3_lh  r3_kills  r3_deaths  r3_items  \
match_id                                                                   
0                5   1909     1143     10         0          0         8   
1                4   1506     1502     18         1          0         7   
2                3   1297      909      0         1          0         6   
3                5   2037     1139     15         0          0         6   
4                3    884      927      0         1          0         8   
...            ...    ...      ...    ...       ...        ...       ...   
114402           3    879     1798     22         1          1         9   
114403           3   1130     1585     20         1          1        14   
114404           3    658      646      2         0          0        10   
114405           2    471      634      2         0          1        10   
114406           3   1103      889      3         1          0         9   

          r4_level  r4_xp  r4_gold  r4_lh  r4_kills  r4_deaths  r4_items  \
match_id                                                                   
0                3    757      741      6         0          0         7   
1                3    669      631      7         0          0         7   
2                5   2360     2096     26         1          1         6   
3                2    591      499      0         0          0         6   
4                3    925     1439     16         1          0        11   
...            ...    ...      ...    ...       ...        ...       ...   
114402           4   1835     1350     11         1          1         9   
114403           4   1546     1166      8         1          0         6   
114404           4   1710     1833     24         1          0        10   
114405           1    265      605      1         0          0         9   
114406           4   1293     1675     11         1          0        11   

          r5_level  r5_xp  r5_gold  r5_lh  r5_kills  r

In [115]:
y = np.array(y)
y = y.ravel()

In [116]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=241)

In [117]:
pca = StandardScaler()
X_train_ = pca.fit_transform(X_train)
X_test_ = pca.fit_transform(X_test)

In [118]:
logreg_num = LogisticRegression(C=c_best, penalty='l2', max_iter=500, random_state=241)
logreg.fit(X_train_, y_train)

LogisticRegression(max_iter=500, random_state=241)

In [119]:
Pred_logreg_test = logreg.predict_proba(X_test_)
Pred_logreg_train = logreg.predict_proba(X_train_)
roc_auc_test = roc_auc_score(y_test, Pred_logreg_test[:,1])
roc_auc_train = roc_auc_score(y_train, Pred_logreg_train[:,1])
print('значение roc_auc на test: ', round(roc_auc_test, 4))
print('значение roc_auc на train: ', round(roc_auc_train, 4))
print('среднее значение roc_auc: ', round((roc_auc_test + roc_auc_train)/2, 4))

значение roc_auc на test:  0.7211
значение roc_auc на train:  0.7163
среднее значение roc_auc:  0.7187


3. Сколько различных идентификаторов героев существует в данной игре?
См. ниже
Всего 112 id

In [120]:
heroes = pd.read_csv(r'S:\Coursera\Введение в машинное обучение\Итоговое задание\data\data\dictionaries\heroes.csv', index_col='id')
heroes

localized_name                      name
id                                           
1         Anti-Mage                  antimage
2               Axe                       axe
3              Bane                      bane
4       Bloodseeker               bloodseeker
5    Crystal Maiden            crystal_maiden
..              ...                       ...
110         Phoenix                   phoenix
105         Techies                   techies
111          Oracle                    oracle
112   Winter Wyvern             winter_wyvern
113      Arc Warden  npc_dota_hero_arc_warden

[112 rows x 2 columns]

4. Какое получилось качество при добавлении "мешка слов" по героям? Улучшилось ли оно по сравнению с предыдущим вариантом? Чем вы можете это объяснить?

Удаление категориальных признаков снижает значение roc_auc с 0.745 до 0.7211. 
Выбор героев значительно влияет на исход игры, поэтому исключение этих признаков из матрицы снижает качество модели.

5. Какое минимальное и максимальное значение прогноза на тестовой выборке получилось у лучшего из алгоритмов?

Результат на каггле 0.75053 - лучшее качество из всех моделей

# ПРЕДСКАЗАНИЯ ДЛЯ ЛОГ. РЕГРЕССИЙ

In [110]:
best_model = logreg_best

In [111]:
X_test_test = pd.read_csv(r'S:\Coursera\Введение в машинное обучение\Итоговое задание\features_test.csv', index_col='match_id')
X_test_test = X_test_test[feature_cols]
X_test_test['first_blood_player1'] = X_test_test['first_blood_player1'].fillna(0)
X_test_test['first_blood_player2'] = X_test_test['first_blood_player2'].fillna(0)
X_test_test['first_blood_team'] = X_test_test['first_blood_team'].fillna(.5)
X_test_test[time_gap_cols] = X_test_test[time_gap_cols].fillna(1000)
X_test_test = cat_to_int(X_test_test, cat_cols)
X_test_test = pd.get_dummies(X_test_test, columns=cat_cols, drop_first=True)
X_test_test_ = pca.fit_transform(X_test_test)
X_test_test_

array([[ 0.47474566, -0.23774298, -0.1370936 , ..., -0.22689511,
        -0.23528192, -0.23181681],
       [-1.36563036, -1.20791332, -1.25566036, ..., -0.22689511,
        -0.23528192, -0.23181681],
       [-1.36563036, -0.86205735, -0.74271453, ..., -0.22689511,
        -0.23528192, -0.23181681],
       ...,
       [-1.36563036, -1.17421453, -1.40868202, ..., -0.22689511,
        -0.23528192, -0.23181681],
       [ 0.47474566,  1.0765097 ,  0.05041181, ..., -0.22689511,
        -0.23528192, -0.23181681],
       [ 0.47474566, -0.0390975 ,  0.32843708, ..., -0.22689511,
        -0.23528192, -0.23181681]])

In [112]:
submission = pd.read_csv(r'S:\Coursera\Введение в машинное обучение\Итоговое задание\submission.csv')
submission['match_id'] = X_test_test.index
submission['radiant_win'] = best_model.predict_proba(X_test_test_)[:,1]

In [113]:
submission.to_csv(r'S:\Coursera\Введение в машинное обучение\Итоговое задание\my_submission_logreg.csv', index=False)

In [123]:
d = pd.DataFrame.from_dict({'1': [1], '2': [2], '3': [3]})
d

1  2  3
0  1  2  3

In [124]:
s = d
del d['2']
s

1  3
0  1  3